File: backend/projects/airtable_mock.py
"""
Mock test double for pyairtable used in unit tests.
Simulates real Airtable API behavior in-memory including rate limit simulation and errors.
"""
import uuid
from typing import Dict, List, Any, Optional


class MockAirtableError(Exception):
    def __init__(self, status_code: int, message: str):
        super().__init__(f"HTTP {status_code}: {message}")
        self.status_code = status_code
        self.message = message


class MockAirtableTable:
    def __init__(self, table_name: str):
        self.table_name = table_name
        self.records: List[Dict[str, Any]] = []
        self.rate_limit_hits_remaining = 0
        self.permanent_error_field: Optional[str] = None
        self.failTaskIds = set()

    def all(self, formula: Optional[str] = None) -> List[Dict[str, Any]]:
        return list(self.records)

    def create(self, fields: Dict[str, Any]) -> Dict[str, Any]:
        if self.rate_limit_hits_remaining > 0:
            self.rate_limit_hits_remaining -= 1
            raise MockAirtableError(429, "Rate limit exceeded. Try again later.")

        task_id = fields.get("Task ID")
        if task_id and task_id in self.failTaskIds:
            raise MockAirtableError(422, f"Unprocessable Entity: invalid task {task_id}")

        if self.permanent_error_field and self.permanent_error_field in fields:
            raise MockAirtableError(422, f"Unprocessable Entity: invalid value for {self.permanent_error_field}")

        rec_id = f"rec{uuid.uuid4().hex[:14]}"
        record = {
            "id": rec_id,
            "fields": dict(fields),
            "createdTime": "2026-09-10T12:00:00.000Z",
        }
        self.records.append(record)
        return record

    def update(self, record_id: str, fields: Dict[str, Any]) -> Dict[str, Any]:
        if self.rate_limit_hits_remaining > 0:
            self.rate_limit_hits_remaining -= 1
            raise MockAirtableError(429, "Rate limit exceeded. Try again later.")

        for r in self.records:
            if r["id"] == record_id:
                r["fields"].update(fields)
                return r
        raise MockAirtableError(404, f"Record {record_id} not found")

    def batch_create(self, records: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        return [self.create(r.get("fields", r)) for r in records]

    def batch_update(self, records: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        results = []
        for r in records:
            results.append(self.update(r["id"], r.get("fields", {})))
        return results

    def clear(self):
        self.records.clear()
        self.rate_limit_hits_remaining = 0
        self.permanent_error_field = None


class MockAirtableBase:
    def __init__(self, base_id: str):
        self.base_id = base_id
        self.tables: Dict[str, MockAirtableTable] = {}

    def table(self, table_name: str) -> MockAirtableTable:
        if table_name not in self.tables:
            self.tables[table_name] = MockAirtableTable(table_name)
        return self.tables[table_name]


class MockAirtableApi:
    def __init__(self, api_key: str = "mock-key"):
        self.api_key = api_key
        self.bases: Dict[str, MockAirtableBase] = {}

    def base(self, base_id: str) -> MockAirtableBase:
        if base_id not in self.bases:
            self.bases[base_id] = MockAirtableBase(base_id)
        return self.bases[base_id]

File: frontend/src/lib/airtable-mock.ts
/**
 * Test double for the Airtable npm SDK.
 * Simulates Airtable in-memory storage, rate limiting, and errors for unit tests.
 */
export type MockAirtableRecord = {
  id: string;
  fields: Record<string, any>;
  createdTime: string;
};

export class MockAirtableTable {
  public records: MockAirtableRecord[] = [];
  public rateLimitCount = 0;
  public failTaskIds: Set<string> = new Set();

  constructor(public tableName: string) {}

  select() {
    return {
      all: async (): Promise<MockAirtableRecord[]> => {
        return [...this.records];
      },
    };
  }

  async create(fields: Record<string, any>): Promise<MockAirtableRecord> {
    if (this.rateLimitCount > 0) {
      this.rateLimitCount--;
      const err: any = new Error("Rate limit exceeded: 429 Too Many Requests");
      err.statusCode = 429;
      throw err;
    }

    const taskId = fields["Task ID"];
    if (taskId && this.failTaskIds.has(taskId)) {
      const err: any = new Error(`Unprocessable Entity: invalid task ${taskId}`);
      err.statusCode = 422;
      throw err;
    }

    const rec: MockAirtableRecord = {
      id: `rec${Math.random().toString(36).substring(2, 16)}`,
      fields: { ...fields },
      createdTime: new Date().toISOString(),
    };
    this.records.push(rec);
    return rec;
  }

  async update(recordId: string, fields: Record<string, any>): Promise<MockAirtableRecord> {
    if (this.rateLimitCount > 0) {
      this.rateLimitCount--;
      const err: any = new Error("Rate limit exceeded: 429 Too Many Requests");
      err.statusCode = 429;
      throw err;
    }

    const idx = this.records.findIndex((r) => r.id === recordId);
    if (idx === -1) {
      const err: any = new Error(`Record ${recordId} not found`);
      err.statusCode = 404;
      throw err;
    }

    this.records[idx].fields = {
      ...this.records[idx].fields,
      ...fields,
    };
    return this.records[idx];
  }
}

export class MockAirtableBase {
  private tables: Map<string, MockAirtableTable> = new Map();

  constructor(public baseId: string) {}

  table(name: string): MockAirtableTable {
    if (!this.tables.has(name)) {
      this.tables.set(name, new MockAirtableTable(name));
    }
    return this.tables.get(name)!;
  }
}

export class MockAirtable {
  private bases: Map<string, MockAirtableBase> = new Map();

  base(baseId: string) {
    if (!this.bases.has(baseId)) {
      this.bases.set(baseId, new MockAirtableBase(baseId));
    }
    return (tableName: string) => this.bases.get(baseId)!.table(tableName);
  }
}
